Adding webhook extension support to the CCM (#2928)
kubernetes/enhancements: keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm/README.md

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP will detail enhancing the CCM framework to support cloud provider
specific webhooks. The intent is to make it easy to either generate a binary
or enhance the existing CCM binary to host such webhooks. We also intend to
allow for easily linking in "standard" webhooks needed by other SIGs which
need to be customized for particular cloud providers.

## Motivation

> The Cloud (Provider) Controller Manager (CCM) is the binary into which the Cloud

Reviewer: I think "cloud controller manager" is used frequently enough: 
```suggestion
The Cloud Controller Manager (CCM) is the binary into which the Cloud
```

Author: Fixed, thanks

Provider places all the controllers needed to make a Kubernetes cluster work
correctly on their Cloud. There are also occasions when it makes sense for a Cloud

> Provider to want these customizations to be applied inline rather than 

Reviewer: ```suggestion
Provider to want these customizations to be applied synchronously, during API server request handling, rather than 
```

Author: Thanks

asynchronously after a change has already been applied. 

Our initial example of this is from SIG Storage. These would like the functionality 

> from PVL admission controller (https://github.com/kubernetes/kubernetes/issues/52617).

Reviewer: ```suggestion
from the PersistentVolumeLabel (PVL) admission controller (https://github.com/kubernetes/kubernetes/issues/52617).
```

Author: Done

This needs to be completed for cloud provider extraction to complete. Several

> Cloud Providers have indicated that this should be done inline, especially as the

Reviewer: I think we should either link to a comment or notes where this is explained, or provide the argument here.  I don't remember why its required to be in-line but I think its an important point to spell out.

Reviewer: Nit: 
```suggestion
This needs to be completed for cloud provider extraction to complete. Several
Cloud Providers have indicated that this should be done in-line, especially as the
```

Author: Done (in several places)

> existing deprecated solution is an inline solution.

Reviewer: ```suggestion
existing deprecated solution is implemented in an API server admission plugin, which is in-line in the request path.
```

Author: Thanks

> ### Goals

Reviewer 2: I'd like to see authn, authz, and serving certificates considered as immediate goals and requirements so that all kinds of admission webhook can use this mechanism.

Author: Added.

> Our imediate goal is to allow in tree Cloud Providers are able to stop using the 

Reviewer 3: Nit: readability improvement, I think these are extra 
```suggestion
Our immediate goal is to allow in tree Cloud Providers to stop using the 
```

Author: Thanks :D

existing PVL admission controller and do so using the framework. However we want to
build a framework which wil be usable by similar solutions to problems. This KEP is
about the framework needed to support the PVL webhook and not the webhook itself.

### Non-Goals

We are not intending to create a [general admission webhook solution](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html). This is just
intended to host Cloud Provider specific webhooks as part of the Control Plane.

## Proposal

We will start by adding extension hooks which can be registered in the 
cmd/cloud-controller-manager/main.go. This would be similar to the mechanism
we already use to register new controllers. The existing sample shows this with
a sample of registering the nodeipamcontroller which is not a normally installed
controller in the cloud controller manager. In a similar way we will have a 
sample of integrating a PVL mutating webhook into the sample CCM. We will also 
have the system automatically detect if there are both controllers and webhooks

> registered in the binary. If both are registered it will automatically add 
> command line flags allowing webhooks and controller to be disabled. The 

Reviewer 3: I assume this would be two separate flags right?
```suggestion
registered in the binary. If both are registered it will automatically add 
command line flags allowing webhooks and/or controllers to be disabled. The 
```

Author: Yes, I talk about both the controllers flag and the webhooks flag in the next few sentences. However its probably better to be explicit, so I've added a sentence to make that explicit.

controller flag will default to the controllers being enabled. The webhook flag
will default to the webhooks being disabled. We would also like to provide a 
builder pattern for registering both the controller and webhook extensions.

> Another issue to consider is how the mutating/asmission webhook configuration is

Reviewer 3: ```suggestion
Another issue to consider is how the mutating/admission webhook configuration is
```

Author: Thanks

written into the cluster. This may be somewhat dependent on if the Cloud Provider
intends to run it on the Control Plane or on the Cluster. We would recommend on 

> running in on the Control Plane. However for some Coud Providers that can lead to

Reviewer 3: This paragraph is becoming hard to follow because we keep referring to the subject as "it". I find in enhancement docs it's better to be explicit, eg `This may be somewhat dependent on if the Cloud Provider
intends to run the webhook on the Control Plane or on the Cluster.`

Aside from that, and maybe this is a lack of context on my part, I'm not sure why there's a distinction here, are you talking about whether they run the webhook on control plane nodes vs worker nodes? Why would that make a difference? Can we expand on that within the Kep somewhere?

Author: I'm usually the one complaining about the problem of using 'it' :smiley_cat: Fixed. 

Added another paragraph with at least some of the concerns specific to running the webhook server in the cluster (in the nodes). I realize there is a little blurring of control plane vs worker and static vs dynamic. I can change if you think it needs to be called out. However at this point I don't want to over complicate this section.

special issues with the configuration. As such we will provide a flag which 

> enables the serviec to automatically register the webhooks as part of startup. 

Reviewer 3: ```suggestion
enables the service to automatically register the webhooks as part of startup. 
```

Author: Fixed :D

However that functionality can be disabled, allowing the Cloud Provider to do 
their own customer registration, as part of cluster setup.

### User Stories (Optional)

> The users of this KEP are Cloud Providers abd feature developers whose code 

Reviewer 3: ```suggestion
The users of this KEP are Cloud Providers and feature developers whose code 
```

Author: Thanks

does not deployment, this is a simple deployment, being fewer processes. It 
does not stop the Cloud Provider from converting to Story 1 later. This system
should make our part of this simple. Obviously the Cloud Provider would have to
change their deployment setup.

#### Story 3 - Immediate Cloud Provider Extraction effort

PVL use case. Cloud Providers want to allow customers to migrate an existing
workload to Kubernetes. That workload uses an existing persistent volume. To
get that workload migrated the end user needs to be able to link the existing
PV into the cluster. However this requires an association which requires calls
out to the cloud provider for certain kinds of storage. Ideally the lookup and
label of the PV to that pre-existing storage happens inline when the PV is 
written. That ensures the write volume is attached to the Node/Pod when it is
scheduled and there are no race conditions.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

Potentially there could be problems running webhooks and controller in the same
process. Delays of 10 seconds or more can cause webhooks to fail. It is important 

> to understand than irrespective of failure mode on the webhook coniguration, 

Reviewer 3: ```suggestion
to understand that irrespective of failure mode on the webhook coniguration, 
```

Author: Done.

timeouts will always turn a webhook call into a FAIL. As such we are making it
easy to easily turn the CCM into two processses to mitigate this. It will be upto
the Cloud Provider to determine if they want the webhook policy to be FAIL or 

> IGNORE. We will default to IGNORE as its the safe option. Incorrectly setting FAIL

Author: default is confusing here. What I meant was the sample. Fixed to reflect that.

can quickly lead to a non functional cluster. Having a FAIL policy on Pods for 
example can prevent the system from allocating the webhook service, which prevents
the webhook from ever passing.

Webhooks are configured by a runtime resource. As a consequence this configuration
can be modified to deleted at runtime. That means that an admin on the cluster can
disable or alter the functionality. This potentially makes it harder for a cloud 
provider to enforce that this logic is being applied. It also means that there 
needs to be a deployment mechanism for the webhook. It is left to the Cloud 
Provider to determine if the need for an inline request is sufficient to override

> these concerns. The Cloud Provider can alternately use the controller route which 

Author: Done

> is not inline or use an actual admission controller, built into the APIServer.

Author: Nice :D

not to use anything from controller runtime. We are saying that if need to choose
between unifying the KCM & CCM code and building with controller runtime, we will
choose unifying the KCM & CCM code base.

## Design Details

A sample  of how the Builder pattern might look is:

```
cmOptions, err := options.NewCloudManagerOptions()
if err != nil {
klog.Fatalf("unable to initialize command options: %v", err)
}
fss := cliflag.NamedFlagSets{}
cloudManagerBuilder := app.NewCloudManagerBuilder("name")
cloudManagerBuilder.setOptions(cmOptions)
cloudManagerBuilder.setFlags(fss)
cloudManagerBuilder.registerWebhook(gvkList, handler)
cloudManagerBuilder.registerWebhook(gvkSecondList, secondHandler)
manager, err := cloudManagerBuilder(wait.NeverStop)
if err != nil {
klog.Fatalf("unable to construct cloud manager: %v", err)
}
err := command.start()
```

This will not alter the existing extension hooks in the controller manager 
framework, as they are critical for backward compatibility. The builders
are meant to be an abstraction layer on top to make the extensions easier to 
use. So for the existing controller manager code you might see changes like:

```

> cloudControllerManagerBuilder.registerController("nodeipamcontroller", handler)

Reviewer 4: Is `handler` a `InitFunc` or an already-initiated `Controller` instance? If it is the latter, I believe the initialization process will need a foundational change.

Author: At this point it would almost certainly be a ControllerInitFuncConstructor. Having said that, this is meant to give a feeling for the builder pattern. It seems likely this will need to be adjusted some when we actually implement. We will of course then take those discoveries and feed them back into the KEP. This is after all still alpha.

### Graduation Criteria

#### Alpha

- Have the sample CCM come up and able to serve PVL mutating webhook.
- Requires that we get a PVL mutating webhook written for at least 1 Cloud Provider.

#### Beta

- TBD

#### GA

- TBD

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Not deprecated

### Upgrade / Downgrade Strategy

- Upgrade is not believed to be an issue at this point.
- Currently we are leaving upgrade as an issue for the Cloud Provider

### Version Skew Strategy

- We are currently assuming that this will be deployed as part of the control
plane. We assume it will be upgraded with the KAS, KCM and CCM.

## Production Readiness Review Questionnaire

- TBD

### Feature Enablement and Rollback

- TBD

###### How can this feature be enabled / disabled in a live cluster?

> This will be built into the CCM by the Cloud Provider. Code must be written

Reviewer 2: As I recall, the CCM has featuregates.  This seems suitable for a featuregate so that consumers will be aware that the flags being set are considered alpha by the library providing them.

Author: Added for tracking purposes.

- Have the sample CCM come up and able to serve PVL mutating webhook.
- Requires that we get a PVL mutating webhook written for at least 1 Cloud Provider.

#### Beta

- TBD

#### GA

- TBD

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Not deprecated

### Upgrade / Downgrade Strategy

- Upgrade is not believed to be an issue at this point.
- Currently we are leaving upgrade as an issue for the Cloud Provider

### Version Skew Strategy

- We are currently assuming that this will be deployed as part of the control
plane. We assume it will be upgraded with the KAS, KCM and CCM.

## Production Readiness Review Questionnaire

- TBD

### Feature Enablement and Rollback

- TBD

###### How can this feature be enabled / disabled in a live cluster?

This will be built into the CCM by the Cloud Provider. Code must be written

> specifically by the Cloud Provider to enable this feature.

Reviewer 5: Are cloud providers expected to use a feature gates to gate this feature? What is the prescription here?

It will control if the listener can be started.
Please note however this is a library and we expect users to vendor this into their own code base.
As such we cannot control if they will remove the check rather than setting the flag.
- Feature gate name: CloudWebhookServer
- Components depending on the feature gate: cloud-controller-manager

###### Does enabling the feature change any default behavior?

This cannot just be "enabled".

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

If you build using our framework, then you will be able to disable using a command line flag.
It can also be disabled by changing the admission webhook configuration.

###### What happens if we reenable the feature if it was previously rolled back?

For new update requests it will work. However it will not change any persisted 
resources, unless they are rewritten.

###### Are there any tests for feature enablement/disablement?

- TBD

### Rollout, Upgrade and Rollback Planning

- TBD

###### How can a rollout or rollback fail? Can it impact already running workloads?

- TBD

###### What specific metrics should inform a rollback?

> - TBD

Reviewer 5: Any webhook request metrics?
